docs/zkapps/tutorials/06-offchain-storage.mdx

```diff
@@ -0,0 +1,401 @@
+
+### Off-Chain Storage and Decentralization
+
+Solutions to storage can be on a spectrum from cheap and more centralized, to more expensive and more decentralized. More decentralized solutions will likely be more expensive due to replicating and proving stored data. 
+
+Given that, we expect there to be different solutions on a continuum, offering multiple options for developers to choose depending on the zkApp they are building and the guarantees they want that zkApp to have.
+
+Some likely things to expect on this spectrum and are under exploration:
+1. A single-server storage solution (this is what is presented here).
+2. A multi-server storage solution that can be run by multiple parties for stronger trust guarantees.
+3. A solution leveraging storage on modular blockchains.
+4. A future hard fork to add purchasable on-chain data storage to Mina.
+5. A future hard fork to add data-storage committees to Mina for horizontally scalable storage.
+
+### Single-Server Off-Chain Storage
+
+This tutorial shares and explains an implementation of the single-server off-chain storage solution mentioned as the first option above. The library provides a REST a server that anyone can run to store data for one or multiple zkApps, as well as a zkApp library to check on-chain if changes have been backed by the server.
+
+This implementation requires a trust assumption for zkApps using it, that both developers and users can trust whoever running the server.
+
+This makes it useful for both prototyping applications that need off-chain storage, and putting applications into production where these trust assumptions are reasonable, but not for zkApps where a trustless solution is needed.
+
+See the library [here](https://github.com/es92/zkApp-offchain-storage) if you'd like to learn more about this implementation and see how it works.
+
+### Further Development of Single-Server Off-Chain Storage
+
+We're looking for a developer to run with the library presented here, improve it, make it more decentralized, and run instances for the community. There is a grant opportunity for taking this on. If interested, please reach out [here](mailto:build@minaprotocol.com).
+
+Some suggested improvements:
+
+* Eliminate DDOS vulnerability by adding a token that limits storage requests.
+* Do not store trees for a smart contract if that contract is misconfigured in a way to prevent cleaning up old data.
+* Add support to the client library for connecting to multiple storage servers, enabling correctness under a majority-honest assumption.
+* Switch the project to a more scalable database implementation (e.g. Redis).
+* Write an implementation for an automatically scalable service (e.g. Cloudflare).
+
+## Implementing a Project Using Off-Chain Storage
+
+As usual, there is sample code for this project, which you can find [here](https://github.com/es92/zkApp-examples/tree/main/06-offchain-storage/contracts).
```

Comment:
Links to specific parts of repos should typically be `y` perma links (press y while viewing a GitHub code file) so that they always work even if a future commit renames or moves files.

The tradeoff is that updates to those files need to be accompanied with updates to the docs rather than coming "for free"